contracts/epochs/FlowEpoch.cdc

```diff
@@ -443,6 +526,259 @@ access(all) contract FlowEpoch {
+                seed: randomSource,
+                startView: startView,
+                endView: endView,
+                stakingEndView: stakingEndView,
+                totalRewards: 0.0,     // will be overwritten in `calculateAndSetRewards` below
+                collectorClusters: [], // will be overwritten in `calculateAndSetRewards` below
+                clusterQCs: [],        // will be overwritten in `calculateAndSetRewards` below
+                dkgKeys: [dkgGroupKey].concat(dkgPubKeys)
+            )
+
+            /// Save the new epoch meta data, it will be referenced as the epoch progresses
+            FlowEpoch.saveEpochMetadata(newEpochMetadata)
+
+            /// Calculate rewards for the current epoch and set the payout for the next epoch
+            // TODO(jord): do we need this? L862 we calculate rewards on enter EpochSetup phase
+            FlowEpoch.calculateAndSetRewards()
+
+            /// Emit the EpochRecover service event.
+            /// This will be processed by the Protocol State, which will then exit EFM
+            /// and enter the recovery epoch at the specified start view.
+            self.emitEpochRecoverEvent(epochCounter: recoveryEpochCounter,
+                startView: startView,
+                stakingEndView: stakingEndView,
+                endView: endView,
+                nodeIDs: nodeIDs,
+                clusterAssignments: clusterAssignments,
+                randomSource: randomSource,
+                targetDuration: targetDuration,
+                targetEndTime: targetEndTime,
+                clusterQCVoteData: clusterQCVoteData,
+                dkgPubKeys: dkgPubKeys,
+                dkgGroupKey: dkgGroupKey,
+                dkgIdMapping: dkgIdMapping,
+            )
+
+            /// Start a new Epoch, which increments the current epoch counter
+            FlowEpoch.startNewEpoch()
+        }
+
+        /// Replaces the currently active epoch with a new "recovery epoch" with the provided configuration.
+        /// This function is used to recover from Epoch Fallback Mode (EFM).
+        /// The "recovery epoch" config will be emitted in the EpochRecover service event.
+        /// The "recovery epoch must have a counter exactly equal to the current epoch counter.
+        /// 
+        /// This function differs from recoverNewEpoch because it replaces the currently active epoch instead of starting a new epoch.
+        /// This function does not calculate or distribute rewards
+        /// Calling recoverCurrentEpoch multiple times does not cause multiple reward payouts.
+        /// Rewards for the recovery epoch will be calculated and paid out during the course of the epoch. 
+        /// CAUTION: This causes data loss by replacing the existing current epoch metadata with the inputs to this function.
+        /// This function exists to recover from potential race conditions that caused a prior recoverCurrentEpoch transaction to fail;
```

Comment:
The main reason this exists is:
- we have a strong requirement for consecutive epoch counters in the Protocol State
- it's possible for an `EpochRecover` service event to be rejected by the Protocol State, because of either human error or timing
  - example of human error: an invalid data field is provided in the recovery transaction
  - example of timing: `firstView` is set correctly, but an epoch extension is added before the `EpochRecover` is processed (see [comment](https://github.com/onflow/flow-core-contracts/pull/440/files#r1811130118))

If an EpochRecover event is rejected, then the smart contract has already stored the data for the recovery epoch, but the Protocol State has not. This function gives us an opportunity to "retry" the recovery without increasing the epoch counter or paying rewards.